src/Tribe/Template.php

```diff
@@ -273,7 +274,164 @@ public function merge_context( $context = array(), $file = null, $name = null )
 	}
 
 	/**
-	 * A very simple method to include a Aggregator Template, allowing filtering and additions using hooks.
+	 * Fetches the path for locating files in the Plugin Folder
+	 *
+	 * @since  TBD
+	 *
+	 * @return string
+	 */
+	protected function get_template_plugin_path() {
+		// Craft the plugin Path
+		$path = array_merge( (array) $this->template_base_path, $this->folder );
+
+		// Implode to avoid Window Problems
+		$path = implode( DIRECTORY_SEPARATOR, $path );
+
+		/**
+		 * Allows filtering of the base path for templates
+		 *
+		 * @since  TBD
+		 *
+		 * @param string $path      Complete path to include the base plugin folder
+		 * @param self   $template  Current instance of the Tribe__Template
+		 */
+		return apply_filters( 'tribe_template_plugin_path', $path, $this );
+	}
+
+	/**
+	 * Fetches the Namespace for the public paths, normaly folders to look for
+	 * in the theme's directory.
+	 *
+	 * @since  TBD
+	 *
+	 * @return array
+	 */
+	protected function get_template_public_namespace() {
+		$namespace = array(
+			'tribe',
+			$this->origin->template_namespace,
+		);
+
+		/**
+		 * Allows filtering of the base path for templates
+		 *
+		 * @since  TBD
+		 *
+		 * @param array  $namespace Which is the namespace we will look for files in the theme
+		 * @param self   $template  Current instance of the Tribe__Template
+		 */
+		return apply_filters( 'tribe_template_public_namespace', $namespace, $this );
+	}
+
+	/**
+	 * Fetches the path for locating files given a base folder normally theme related
+	 *
+	 * @since  TBD
+	 *
+	 * @param  mixed  $base  Base path to look into
+	 *
+	 * @return string
+	 */
+	protected function get_template_public_path( $base ) {
+		// Craft the plugin Path
+		$path = array_merge( (array) $base, (array) $this->get_template_public_namespace() );
+
+		// Implode to avoid Window Problems
+		$path = implode( DIRECTORY_SEPARATOR, $path );
+
+		/**
+		 * Allows filtering of the base path for templates
+		 *
+		 * @since  TBD
+		 *
+		 * @param string $path      Complete path to include the base public folder
+		 * @param self   $template  Current instance of the Tribe__Template
+		 */
+		return apply_filters( 'tribe_template_public_path', $path, $this );
+	}
+
+	/**
+	 * Fetches the folders in which we will look for a given file
+	 *
+	 * @since  TBD
+	 *
+	 * @return array
+	 */
+	protected function get_template_path_stack() {
+		$stack = array();
+
+		// Only look into public folders if we tell to use stack
+		if ( $this->use_stack ) {
```

Comment:
Just few thoughts on this one in particular as the overall PR seems great but wanted to have your comments on the following items.

- Name of the structure, in this particular case is referred as stack, however the behavior of the structure is the one from a Queue as it's FIFO (First in first out) so I think the name of stack is great however I think the operations should be LIFO (Last in first out). 

- Data structure I think using an array to represent the stack will give a bit of flexibility instead like: 

```php
$stack[] = array(
	'type' => 'plugin',
	'path' => '',
);
```

- In terms of the theme / child theme I think a nested array structure have a bit more sense, however I'm not 100% confident as will introduce complexity during the iteration of the loop, this is what I mean.


```php
$stack[] = array(
	'type' => 'theme'
	'path' => '',
	'child' => array(
		'type' => 'theme',
		'path' => '',
	)
)
```

So during the loop of the stack if is a theme and has a child the child should take priority, I think this one gives a clear sense of nested structure and dependency on the themes.

Reply:
Yeah I imagined it would be a good idea to tackle and move into a more args like structure for the Stack.

Will address the naming and the how we store it. As of the Child, I will avoid it because it adds complexity we dont need.